Added `zend_simd.h`

SakiTakamachi opened the pull request:
For SIMD using 128-bit vectors, add a header file to use common code between SSE2 and ARM Neon.

I don't intend to replace everything with these.
The goal is to avoid having to copy a lot of "almost identical code" for non-SIMD code on ARM.

There are other places could use it, but it would involve too many changes, so I'll address that in a follow-up PR.

dstogov reviewed (commented):
I don't have a strong opinion about this.

On one hand this enables NEON code on AArch64 almost for free.
On the other hand this requires usage of our own undocumented abstraction layer.
Most probably this layer is more limited than SSE or NEON by their selves.

Personally, I would prefer something like NEON2SSE (e.g. https://github.com/intel/ARM_NEON_2_x86_SSE)

SakiTakamachi commented:
@dstogov 

Thank you for checking!

There isn’t much SIMD code in php-src yet, so for now I chose not to use any external libraries.
However, as you mentioned, using a library would provide greater flexibility, and I have absolutely no problem switching to a library if needed.

Since the SIMD code currently in php-src is based on SSE, NEON2SSE might not fit our purpose, as the direction is reversed.

We probably need something like sse2neon:
https://github.com/DLTcollab/sse2neon

Or alternatively:
https://github.com/simd-everywhere/simde

What do you think about these libraries?

dstogov commented:
> Since the SIMD code currently in php-src is based on SSE, NEON2SSE might not fit our purpose, as the direction is reversed.

OK I didn't look too deep. Anyway, all these libraries are more complex then we really need. It may be simpler to create our own neon<-->sse wrapper that implements SSE intrinsics through NEON instructions. Similar to the API you proposed, but without introducing new "naming". This way the subset of supported SSE code should start to work on NEON out of the box.

Finally, it may be better to duplicate code with native NEON optimization instead of trying reusing SSE optimized code.

These are just my thoughts. I'm not sure what solution is the best.

SakiTakamachi commented:
@dstogov 

> It may be simpler to create our own neon<-->sse wrapper that implements SSE intrinsics through NEON instructions. Similar to the API you proposed, but without introducing new "naming". This way the subset of supported SSE code should start to work on NEON out of the box.

Sounds simple and good.
I’ll try to revise it based on that idea!

> Finally, it may be better to duplicate code with native NEON optimization instead of trying reusing SSE optimized code.

If the differences in code due to optimization are likely to be significant, I think it’s better to write them separately.
If the differences are small, a wrapper might be sufficient.

SakiTakamachi commented:
Since bcmath had its own `simd.h`, I reverted those changes and modified it to use `zend_simd.h` instead.

As for other parts, I updated them where the current definitions in `zend_simd.h` are sufficient for now.

SakiTakamachi commented:
@dstogov 
How about this naming?

bukka commented on Zend/zend_simd.h:
```diff
@@ -0,0 +1,83 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Saki Takamachi <saki@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
```
I think it would be better to license it under BSD (there are already other files) and not give copyright to Zend Technologies.

bukka replied:
Although copyright probably doesn't matter so  I don't really mind. It's kind of the same like giving it to PHP Group. Once PHP Foundation will become legal entity, this should be much easier...

bukka replied:
Maybe add you as a copy right holder and just change it to BSD. It would be good because it will make easier to use this file elsewhere without pulling in Zend license...

bukka replied:
By elsewhere I mean outside php-src...

bukka replied:
I would like to potentially pull it to jso (project that I currently use for JsonSchema development). So it might be actually even better to license it under MIT.

dstogov reviewed (commented):
I think we shouldn't modify ``libbcmath`` and make it dependent on Zend includes.
The rest looks fine.

SakiTakamachi commented:
I created a library named "XSSE" under the MIT license, and I modified it to be used as `zend_simd.h`.
https://github.com/SakiTakamachi/xsse

dstogov reviewed (approved):
I would think about moving ``zend_simd.h`` out of ``Zend`` (it's not under Zend license) and may be renaming (as you are going to reuse this in other projects e.g. into ``xsse.h``  or ``emmintrin_neon.h``). I would also recommend to add a short comment at start of ``zend_simd.h`` (a couple of lines e.g. a wrapper implementing ``<emmintrin.h>`` API for ARM/NEON)

I don't see any technical problems. My recommendations are optional.
